internal/graph/membershipset.go

```diff
@@ -0,0 +1,254 @@
+	ms := NewMembershipSet()
+	for resourceID, result := range mp {
+		ms.addMember(resourceID, result)
+	}
+	return ms
+}
+
+// MembershipSet is a helper set that tracks the membership results for a dispatched Check
+// request, including tracking of the caveats associated with found resource IDs.
+type MembershipSet struct {
+	membersByID         map[string]*v1.CaveatExpression
+	hasDeterminedMember bool
+}
+
+// AddDirectMember adds a resource ID that was *directly* found for the dispatched check, with
+// optional caveat found on the relationship.
+func (ms *MembershipSet) AddDirectMember(resourceID string, caveat *core.ContextualizedCaveat) {
+	ms.addMember(resourceID, wrapCaveat(caveat))
+}
+
+// AddMemberViaRelationship adds a resource ID that was found via another relationship, such
+// as the result of an arrow operation. The `parentRelationship` is the relationship that was
+// followed before the resource itself was resolved. This method will properly apply the caveat(s)
+// from both the parent relationship and the resource's result itself, assuming either have a caveat
+// associated.
+func (ms *MembershipSet) AddMemberViaRelationship(
+	resourceID string,
+	resourceCaveatExpression *v1.CaveatExpression,
+	parentRelationship *core.RelationTuple,
+) {
+	intersection := caveatAnd(wrapCaveat(parentRelationship.Caveat), resourceCaveatExpression)
+	ms.addMember(resourceID, intersection)
+}
+
+func (ms *MembershipSet) addMember(resourceID string, caveatExpr *v1.CaveatExpression) {
+	existing, ok := ms.membersByID[resourceID]
+	if !ok {
+		ms.hasDeterminedMember = ms.hasDeterminedMember || caveatExpr == nil
+		ms.membersByID[resourceID] = caveatExpr
+		return
+	}
+
+	// If a determined membership result has already been found (i.e. there is no caveat),
+	// then nothing more to do.
+	if existing == nil {
+		return
+	}
+
+	// If the new caveat expression is nil, then we are adding a determined result.
+	if caveatExpr == nil {
+		ms.hasDeterminedMember = true
+		ms.membersByID[resourceID] = nil
+		return
+	}
+
+	// Otherwise, the caveats get unioned together.
+	ms.membersByID[resourceID] = caveatOr(existing, caveatExpr)
+}
+
+// UnionWith combines the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) UnionWith(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		ms.addMember(resourceID, details.Expression)
+	}
+}
+
+// IntersectWith intersects the results found in the given map with the members of this set.
+// The changes are made in-place.
+func (ms *MembershipSet) IntersectWith(resultsMap CheckResultsMap) {
+	for resourceID, details := range resultsMap {
+		existing, ok := ms.membersByID[resourceID]
+		if !ok {
+			continue
+		}
+
+		ms.membersByID[resourceID] = caveatAnd(existing, details.Expression)
+	}
+
+	hasDeterminedMember := false
+	for resourceID, expr := range ms.membersByID {
```

Comment:
what if we move this block up, just handling the deletions of the items that do not exist in the input map at the beginning of this method, and do the  loop over `resultsMap`: we can handle the `hasDeterminedMember` logic in the same if block, so you can do optimizations like not calling `caveatAnd`, just like you did in `addMember`. As it stands, one has to bear with the cognitive load that `caveatAnd` has some optimizations that would equal to "noop" assignment to the map.

```go
func (ms *MembershipSet) IntersectWith(resultsMap CheckResultsMap) {
	for resourceID, _ := range ms.membersByID {
		if _, ok := resultsMap[resourceID]; !ok {
			delete(ms.membersByID, resourceID)
		}
	}

	ms.hasDeterminedMember = false
	for resourceID, details := range resultsMap {
		existing, ok := ms.membersByID[resourceID]
		if !ok {
			continue
		}
		if existing == nil && details.Expression == nil {
			ms.hasDeterminedMember = true
			continue
		}

		ms.membersByID[resourceID] = caveatAnd(existing, details.Expression)
	}
}
```